subsys/bluetooth/host/cs.c

```diff
@@ -22,6 +22,189 @@ LOG_MODULE_REGISTER(bt_cs);
+	net_buf_slist_put(&reassembly_bufs, buf);
+
+	LOG_DBG("Allocated new reassembly buffer for conn handle %d", conn_handle);
+	return buf;
+}
+
+/** @brief Frees a reassembly buffer
+ *
+ * @note Takes the ownership of the pointer and sets it to NULL
+ *
+ * @param buf Double pointer to reassembly buffer
+ */
+static void free_reassembly_buf(struct net_buf **buf)
+{
+	if (!buf) {
+		LOG_ERR("NUll double pointer was passed when attempting to free reassembly buffer");
+		return;
+	}
+
+	if (!(*buf)) {
+		LOG_WRN("Attempted double free on reassembly buffer");
+		return;
+	}
+
+	struct reassembly_buf_meta_data *buf_meta_data =
+		(struct reassembly_buf_meta_data *)((*buf)->user_data);
+
+	LOG_DBG("De-allocating reassembly buffer for conn handle %d", buf_meta_data->conn_handle);
+	if (!sys_slist_find_and_remove(&reassembly_bufs, &(*buf)->node)) {
+		LOG_WRN("The buffer was not in the list");
+	}
+
+	net_buf_unref(*buf);
+	*buf = NULL;
+}
+
+/** @brief Gets the reassembly buffer identified by the connection handle
+ *
+ * @param conn_handle Connection handle
+ * @param allocate Allocates a new reassembly buffer if it's not allocated already
+ * @return struct net_buf* Reassembly buffer, NULL if it doesn't exist or failed when allocating new
+ */
+static struct net_buf *get_reassembly_buf(uint16_t conn_handle, bool allocate)
+{
+	sys_snode_t *node;
+
+	SYS_SLIST_FOR_EACH_NODE(&reassembly_bufs, node) {
+		struct net_buf *buf = CONTAINER_OF(node, struct net_buf, node);
+		struct reassembly_buf_meta_data *buf_meta_data =
+			(struct reassembly_buf_meta_data *)(buf->user_data);
+
+		if (buf_meta_data->conn_handle == conn_handle) {
+			return buf;
+		}
+	}
+
+	return allocate ? alloc_reassembly_buf(conn_handle) : NULL;
+}
+
+/** @brief Adds step data to a reassembly buffer
+ *
+ * @param reassembly_buf Reassembly buffer
+ * @param data Step data
+ * @param data_len Step data length
+ * @return true if successful, false if there is insufficient space
+ */
+static bool add_reassembly_data(struct net_buf *reassembly_buf, const uint8_t *data,
+				uint16_t data_len)
+{
+	if (data_len > net_buf_tailroom(reassembly_buf)) {
+		LOG_ERR("Not enough reassembly buffer space for subevent result");
+		return false;
+	}
+
+	net_buf_add_mem(reassembly_buf, data, data_len);
+	return true;
+}
+
+/** @brief Initializes a reassembly buffer from partial step data
+ *
+ * @param conn_handle Connection handle
+ * @param steps Step data
+ * @param step_data_len Step data length
+ * @return struct net_buf* Pointer to reassembly buffer, NULL if fails to allocate or insert data
+ */
+static struct net_buf *start_reassembly(uint16_t conn_handle, const uint8_t *steps,
+					uint16_t step_data_len)
+{
+	struct net_buf *reassembly_buf = get_reassembly_buf(conn_handle, true);
+
+	if (!reassembly_buf) {
+		LOG_ERR("No buffer allocated for the result reassembly");
+		return NULL;
+	}
+
+	if (reassembly_buf->len) {
+		LOG_WRN("Over-written incomplete CS subevent results");
```

Comment:
In cases like this should we raise the callback to the app for the previous result? Or is this not expected to happen

Reply:
Normally, we allow partial results when the cs procedure or subevent is aborted. But in that case, the controller reports this by setting the subevent and procedure done status bits accordingly.

What we detect here is an abnormal use case where the controller sends some partial results followed by a new subevent results without marking the previous ones as done/aborted. 

I'm not sure reporting this data is necessary because there is no way to verify its integrity.